distributed/spans.py

```diff
@@ -205,72 +266,92 @@ def nbytes_total(self) -> int:
+
+    #: All spans, keyed by the individual tags that make up their name and sorted by
+    #: creation time.
+    #: This is a convenience helper structure to speed up searches.
+    spans_search_by_tag: defaultdict[str, list[Span]]
 
     def __init__(self, scheduler: Scheduler):
         self.spans = {}
-        self.root_spans = {}
-        self.spans_search_by_tag = defaultdict(set)
+        self.root_spans = []
+        self.spans_search_by_name = defaultdict(list)
+        self.spans_search_by_tag = defaultdict(list)
 
-    def new_tasks(self, tss: Iterable[TaskState]) -> dict[str, tuple[str, ...]]:
+    def new_tasks(self, tss: Iterable[TaskState]) -> None:
         """Acknowledge the creation of new tasks on the scheduler.
         Attach tasks to either the desired span or to ("default", ).
-        Update TaskState.annotations["span"] and TaskGroup.span.
-
-        Returns
-        -------
-        {task key: span id}, only for tasks that explicitly define a span
+        Update TaskGroup.span_id and wipe TaskState.annotations["span"].
         """
-        out = {}
+        default_span = None
+
         for ts in tss:
             # You may have different tasks belonging to the same TaskGroup but to
             # different spans. If that happens, arbitrarily force everything onto the
             # span of the earliest encountered TaskGroup.
             tg = ts.group
-            if tg.span:
-                span_id = tg.span
-            else:
-                span_id = ts.annotations.get("span", ("default",))
-                assert isinstance(span_id, tuple)
-                tg.span = span_id
-                span = self._ensure_span(span_id)
+            if not tg.span_id:
+                ann = ts.annotations.get("span")
+                if ann:
+                    span = self._ensure_span(ann["name"], ann["ids"])
+                else:
+                    if not default_span:
+                        default_span = self._ensure_default_span()
+                    span = default_span
+
+                tg.span_id = span.id
                 span.groups.add(tg)
 
-            # Override ts.annotations["span"] with span_id from task group
-            if span_id == ("default",):
-                ts.annotations.pop("span", None)
-            else:
-                ts.annotations["span"] = out[ts.key] = span_id
-
-        return out
+            # The span may be completely different from the one referenced by the
+            # annotation, due to the TaskGroup collision issue explained above.
+            # Remove the annotation to avoid confusion, and instead rely on
+            # distributed.scheduler.TaskState.group.span_id and
+            # distributed.worker_state_machine.TaskState.span_id.
+            ts.annotations.pop("span", None)
+
+    def _ensure_default_span(self) -> Span:
+        """Return the currently active default span, or create one if the previous one
```

Comment:
Out-of-scope: I'm not a huge fan of this as it's not very reproducible. Let's keep thinking about a better way to reproduce the progress/code-tracking behavior from computations.

Reply:
I don't know; I don't hate it now.
If you don't explicitly use spans, you get something that somehow makes sense, at the cost of some mild non-determinism.
If you do use spans, everything is deterministic.